sky/shell/platform/android/platform_view_android.cc

```diff
@@ -38,51 +258,71 @@ bool PlatformViewAndroid::Register(JNIEnv* env) {
   return RegisterNativesImpl(env);
 }
 
-PlatformView* PlatformView::Create(const Config& config) {
-  return new PlatformViewAndroid(config);
+// Per platform implementation of PlatformView::Create
+PlatformView* PlatformView::Create(const Config& config,
+                                   SurfaceConfig surface_config) {
+  return new PlatformViewAndroid(config, surface_config);
 }
 
-PlatformViewAndroid::PlatformViewAndroid(const Config& config)
-  : PlatformView(config), window_(nullptr), did_draw_(false, false) {
-}
+PlatformViewAndroid::PlatformViewAndroid(const Config& config,
+                                         SurfaceConfig surface_config)
+    : PlatformView(config, surface_config), weak_factory_(this) {}
 
 PlatformViewAndroid::~PlatformViewAndroid() {
-  if (window_)
-    ReleaseWindow();
+  weak_factory_.InvalidateWeakPtrs();
+}
+
+void PlatformViewAndroid::SetShellView(std::unique_ptr<ShellView> shell_view) {
+  DCHECK(!shell_view_);
+  shell_view_ = std::move(shell_view);
 }
 
 void PlatformViewAndroid::Detach(JNIEnv* env, jobject obj) {
   shell_view_.reset();
   // Note: |this| has been destroyed at this point.
 }
 
-void PlatformViewAndroid::SurfaceCreated(JNIEnv* env, jobject obj, jobject jsurface) {
+void PlatformViewAndroid::SurfaceCreated(JNIEnv* env,
+                                         jobject obj,
+                                         jobject jsurface) {
   base::android::ScopedJavaLocalRef<jobject> protector(env, jsurface);
   // Note: This ensures that any local references used by
   // ANativeWindow_fromSurface are released immediately. This is needed as a
   // workaround for https://code.google.com/p/android/issues/detail?id=68174
   {
     base::android::ScopedJavaLocalFrame scoped_local_reference_frame(env);
-    window_ = ANativeWindow_fromSurface(env, jsurface);
+    ANativeWindow* window = ANativeWindow_fromSurface(env, jsurface);
+    auto context = WTF::MakeUnique<AndroidGLContext>(window, surface_config_);
+    if (context->IsValid()) {
+      DLOG(INFO) << "Could not create a valid window context.";
+      context_ = std::move(context);
+    }
+    ANativeWindow_release(window);
   }
-  SurfaceNotificationsDirect::NotifyCreated(config_, window_, &did_draw_);
-  did_draw_.Wait();
+
+  NotifyCreated();
 }
 
 void PlatformViewAndroid::SurfaceDestroyed(JNIEnv* env, jobject obj) {
-  DCHECK(window_);
-  SurfaceNotificationsDirect::NotifyDestroyed(config_);
-  ReleaseWindow();
+  context_ = nullptr;
```

Comment:
Shouldn't we null this out after calling NotifyDestroyed so we're not racing the null write in SwapBuffers ?